pkg/ldap/datasource.go

```diff
@@ -23,12 +23,243 @@ import (
+		)
+	}
+
+	ldapResp := proxyResp.GetLdapSearchResponse()
+	if ldapResp == nil {
+
+		return nil, &framework.Error{
+			Message: "Error received nil response from the proxy",
+			Code:    api_adapter_v1.ErrorCode_ERROR_CODE_INTERNAL,
+		}
+	}
+
+	if err = json.Unmarshal([]byte(ldapResp.Response), response); err != nil {
+
+		return nil, &framework.Error{
+			Message: fmt.Sprintf("Error unmarshalling ldap response from the proxy, %v.", err),
+			Code:    api_adapter_v1.ErrorCode_ERROR_CODE_INTERNAL,
+		}
+	}
+
+	return response, nil
+}
+
+// Request sends a paginated search query directly to an LDAP server and returns the processed response.
+func (c *ldapClient) Request(_ context.Context, request *Request) (*Response, *framework.Error) {
+	tlsConfig, sgnlErr := GetTLSConfig(request)
+	if sgnlErr != nil {
+		return nil, sgnlErr
+	}
+
+	conn, err := ldap_v3.DialURL(request.BaseURL, ldap_v3.DialWithTLSConfig(tlsConfig))
+	if err != nil {
+		return nil, &framework.Error{
+			Message: fmt.Sprintf("Failed to dial ldap server - %v", err),
+			Code:    api_adapter_v1.ErrorCode_ERROR_CODE_INTERNAL,
+		}
+	}
+
+	// Bind credentials with connection
+	err = conn.Bind(request.BindDN, request.BindPassword)
+	if err != nil {
+		return nil, &framework.Error{
+			Message: fmt.Sprintf("Failed to bind credentials - %v", err),
+			Code:    api_adapter_v1.ErrorCode_ERROR_CODE_DATASOURCE_AUTHENTICATION_FAILED,
+		}
+	}
+	defer conn.Close()
+
+	// Set cursor from the request (if exists)
+	pageControl, ferr := SetPageControl(request)
+	if err != nil {
+		return nil, ferr
+	}
+
+	filters, filterErr := SetFilters(request)
+	if filterErr != nil {
+		return nil, filterErr
+	}
+
+	attributes := []string{}
+
+	for _, attr := range request.Attributes {
+		attributes = append(attributes, attr.ExternalId)
+	}
+
+	// Define LDAP search with filtering, attributes and paging.
+	searchRequest := ldap_v3.NewSearchRequest(
+		request.BaseDN,                 // BaseDN
+		ldap_v3.ScopeWholeSubtree,      // Scope
+		ldap_v3.DerefAlways,            // DeferAliases
+		0,                              // SizeLimit
+		request.RequestTimeoutSeconds,  // TimeLimit
+		false,                          // TypesOnly
+		filters,                        // Filters
+		attributes,                     // Attributes
+		[]ldap_v3.Control{pageControl}, // Controls
+	)
+
+	response := &Response{}
+
+	// Perform search
+	searchResult, err := conn.Search(searchRequest)
+	if err != nil {
+		// Extract LDAP result code from the error
+		if ldapErr, ok := err.(*ldap_v3.Error); ok {
+			statusCode := ResultCodeToHTTPStatusCode(ldapErr)
+			response.StatusCode = statusCode
+
+			return response, nil
+		}
+
+		return nil, customerror.UpdateError(&framework.Error{
+			Message: fmt.Sprintf("Error searching LDAP server: - %v.", err),
+			Code:    api_adapter_v1.ErrorCode_ERROR_CODE_INTERNAL,
+		},
+			customerror.WithRequestTimeoutMessage(err, request.RequestTimeoutSeconds),
+		)
```

Comment:
nit: formatting + remove dash
```suggestion
		return nil, customerror.UpdateError(
			&framework.Error{
				Message: fmt.Sprintf(
					"Error searching LDAP server: %v.", err,
				),
				Code:    api_adapter_v1.ErrorCode_ERROR_CODE_INTERNAL,
			},
			customerror.WithRequestTimeoutMessage(
				err, request.RequestTimeoutSeconds,
			),
		)
```